### PR TITLE
Move capture lowering from THIR to MIR

### LINKED ISSUE #25: Move Upvar lowering from THIR to MIR
We lower paths that start from an `Upvar` [here](https://github.com/sexxi-goose/rust/blob/f4d014cee77f9112b2582d86085631c74413eb1e/compiler/rustc_mir_build/src/thir/cx/expr.rs#L909-L1006). 

Essentially what this code does is convert upvars to refer to a field within the desugared `ClosureSubsts`. We want to refactor this code as follows:
- Instead of doing the above desugaring we want to return `Thir::Expr::UpvarRef(var_hir_id)` in case the var_hir_id is mentioned within the closure. (This will allow us to deal with the `let _ = x` problem in MIR).
- Move the logic currently in `convert_var` to [`rustc_mir_build/as_place`](https://github.com/sexxi-goose/rust/blob/f4d014cee77f9112b2582d86085631c74413eb1e/compiler/rustc_mir_build/src/build/expr/as_place.rs#L122-L129)
- Maybe remove `Thir::Expr::SelfRef`, it's most likely only for Closures.

### PR DESCRIPTION
**Pull request by arora-aman**
This allows us to:
- Handle precise Places captured by a closure directly in MIR. Handling
  captures in MIR is easier since we can rely on/ tweak PlaceBuilder to
  generate `mir::Place`s that resemble how we store captures (`hir::Place`).

- Handle `let _ = x` case when feature `capture_disjoint_fields`
  is enabled directly in MIR. This is required to be done in MIR since
  patterns are desugared in MIR.

Closes: rust-lang/project-rfc-2229#25

---
This change is [Reviewable](https://reviewable.io/reviews/sexxi-goose/rust/30)

### CODE REVIEW

File: compiler/rustc_mir_build/src/build/expr/as_place.rs [hunk 1]
```diff
@@ -270,6 +294,61 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
...
+        let substs = match closure_ty.kind() {
+            ty::Closure(_, substs) => ty::UpvarSubsts::Closure(substs),
+            ty::Generator(_, substs, _) => ty::UpvarSubsts::Generator(substs),
+            _ => bug!("Lowering capture for non-closure type {:?}", closure_ty)
+        };
+
+        // Access the capture by accessing the field within the Closure struct.
+        //
+        // We must have inferred the capture types since we are build MIR, therefore
+        // it's safe to call `upvar_tys` and we know can unwrap here because
```
Comment (roxelo) [inline compiler/rustc_mir_build/src/build/expr/as_place.rs]: now

File: compiler/rustc_mir_build/src/build/expr/as_place.rs [hunk 2]
```diff
@@ -270,6 +294,61 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
...
+
+        let substs = match closure_ty.kind() {
+            ty::Closure(_, substs) => ty::UpvarSubsts::Closure(substs),
+            ty::Generator(_, substs, _) => ty::UpvarSubsts::Generator(substs),
+            _ => bug!("Lowering capture for non-closure type {:?}", closure_ty)
+        };
+
+        // Access the capture by accessing the field within the Closure struct.
+        //
+        // We must have inferred the capture types since we are build MIR, therefore
```
Comment (roxelo) [inline compiler/rustc_mir_build/src/build/expr/as_place.rs]: building

**Review by roxelo — APPROVED** [review]
Some nits but otherwise lgtm

**Comment by arora-aman** [comment]
Fixed and created rust-lang PR here: https://github.com/rust-lang/rust/pull/79149

**Pull request closed without merging**